modules/auxiliary/gather/ldap_object_attribute.rb

```diff
@@ -0,0 +1,147 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Remote::LDAP
+  include Msf::OptionalSession::LDAP
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'LDAP Update Object',
+        'Description' => %q{
+          This module allows creating, reading, updating and deleting attributes of LDAP objects.
+          Users can specify the object and must specify a corresponding attribute.
+        },
+        'Author' => ['jheysel'],
+        'License' => MSF_LICENSE,
+        'Actions' => [
+          ['CREATE', { 'Description' => 'Create an LDAP object' }],
+          ['READ', { 'Description' => 'Read the the LDAP object' }],
+          ['UPDATE', { 'Description' => 'Modify the LDAP object' }],
+          ['DELETE', { 'Description' => 'Delete the LDAP object' }]
+        ],
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [],
+          'SideEffects' => [IOC_IN_LOGS, CONFIG_CHANGES]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptEnum.new('OBJECT_LOOKUP', [true, 'How to look up the target LDAP object', 'dN', ['dN', 'sAMAccountName']]),
+        OptString.new('OBJECT', [true, 'The target LDAP object']),
+        OptString.new('ATTRIBUTE', [true, 'The LDAP attribute to update (e.g., userPrincipalName)']),
+        OptString.new('VALUE', [true, 'The value for the specified LDAP object attribute'], conditions: ['ACTION', 'in', %w[Create Update] ])
+      ]
+    )
+  end
+
+  def find_target_object
+    search_filter = "(&(#{ldap_escape_filter(datastore['OBJECT_LOOKUP'])}=#{ldap_escape_filter(datastore['OBJECT'])}))"
+    result = []
+
+    @ldap.search(base: @base_dn, filter: search_filter, attributes: ['distinguishedName', datastore['ATTRIBUTE']]) do |entry|
+      result << entry
+    end
```

Comment:
I wasn't able to get this to work when searching by DN. I made this tweak to set the `base_dn` for the query and use the left most component as the search filter which should work.

I also dropped the attribute, AFAIK you always get the object ID in the `dN` and that's more generically compatible with LDAP servers but references to `distinguishedName` will need to be updated to `dN`.

```suggestion
    if datastore['OBJECT_LOOKUP'] == 'dN'
      search_filter, _, base_dn = datastore['OBJECT'].partition(',')
      search_scope = Net::LDAP::SearchScope_SingleLevel
    else
      search_filter = "(#{ldap_escape_filter(datastore['OBJECT_LOOKUP'])}=#{ldap_escape_filter(datastore['OBJECT'])})"
      base_dn = @base_dn
      search_scope = Net::LDAP::SearchScope_WholeSubtree
    end

    result = []

    @ldap.search(base: base_dn, filter: search_filter, attributes: [datastore['ATTRIBUTE']], scope: search_scope) do |entry|
      result << entry
    end
```